app/ota_client.py

```diff
@@ -50,73 +49,87 @@ def verify_file(filename: Path, filehash: str, filesize) -> bool:
-        error_count += 1
-        do_retry = True
-        if isinstance(e, requests.exceptions.Timeout):
-            last_error = f"requests timeout: {e},{url=} ({error_count})"
-        elif isinstance(e, requests.exceptions.ConnectionError):
-            last_error = f"requests connection error: {e},{url=} ({error_count})"
-        elif isinstance(e, requests.exceptions.ChunkedEncodingError):
-            last_error = f"requests ChunkedEncodingError: {url=} ({error_count})"
-        elif isinstance(e, RequestException):
-            status_code = getattr(e.response, "status_code", None)
-            last_error = f"requests error: {status_code=},{url=} ({error_count})"
-            if isinstance(status_code, int) and status_code // 100 == 4:
-                do_retry = False
-        else:
-            last_error = f"requests error unknown: {e},{url=}, ({error_count})"
-        logger.warning(last_error)
-        return do_retry
-
-    @retry(
-        stop_max_attempt_number=5,
-        retry_on_exception=retry_if_possible,
-        wait_exponential_multiplier=1_000,
-        wait_exponential_max=10_000,
-    )
-    def _requests_get():
-        CHUNK_SIZE = 4096
-        headers = {}
-        headers["Accept-encording"] = "gzip"
-        response = requests.get(
-            url, headers=headers, cookies=cookies, timeout=10, stream=True
-        )
-        response.raise_for_status()
+class Downloader:
+    CHUNK_SIZE = 2 * 1024 * 1024  # 2MiB
+    RETRY_COUNT = 5
 
-        with open(dst, "wb") as f:
-            m = sha256()
-            total_length = response.headers.get("content-length")
-            if total_length is None:
-                m.update(response.content)
-                f.write(response.content)
-            else:
-                for data in response.iter_content(chunk_size=CHUNK_SIZE):
-                    m.update(data)
-                    f.write(data)
-        return m.hexdigest()
+    def __init__(self, *, proxy: str):
+        from requests.adapters import HTTPAdapter
+        from urllib3.util.retry import Retry
+
+        # base session
+        session = requests.Session()
 
-    try:
-        calc_digest = _requests_get()
-    except Exception:
-        logger.exception("requests_get")
-        raise OtaErrorRecoverable(last_error)
+        # configure proxy
+        proxies = {"http": proxy, "https": ""}
+        session.proxies.update(proxies)
 
-    if digest and digest != calc_digest:
-        raise OtaErrorRecoverable(
-            f"hash error: act={calc_digest}, exp={digest}, {url=}"
+        # init retry mechanism
+        retry_strategy = Retry(
+            total=self.RETRY_COUNT,
+            raise_on_status=True,
+            backoff_factor=1,
+            # retry on common server side errors and non-critical client side errors
+            status_forcelist={413, 429, 500, 502, 503, 504},
+            allowed_methods=["GET"],
         )
+        retry_strategy.DEFAULT_BACKOFF_MAX = 10
```

Comment:
it should be fixed as:
```suggestion
        Retry.DEFAULT_BACKOFF_MAX = 10
```

Reply:
Emmmm, `retry_strategy` is an instance of `Retry`, after the instance is created and initialized, changing the original class' `DEFAULT_BACKOFF_MAX` will not impact the corresponding field in the `retry_strategy`.

```python
>>> class A:
...     E = 123
... 
>>> a = A()
>>> a.E
123
>>> a.E = 456
>>> a.E
456
>>> A.E
123
>>> A.E = 789
>>> A.E
789
>>> a.E
456
>>> b = A()
>>> b.E
789
```

Reply:
By the way, start from urllib3 ver 2.0, the DEFAULT_BACKOFF_MAX can be configured via __init__, should we upgrade above v2.0?

Reply:
Did you confirmed `retry_strategy.DEFAULT_BACKOFF_MAX = 10` is working properly?
My snippet did not work. Retry.DEFAULT_BACKOFF_MAX = 10 works.

Reply:
It depends on the version of urllib3, as for version 1.26.8, the logic for deciding the backoff value is as follow
`return min(self.DEFAULT_BACKOFF_MAX, backoff_value)`, in this case, `retry_strategy.DEFAULT_BACKOFF_MAX = 10` should work.

Reply:
Did you confirm that is working? If so I think it is OK :+1:

Reply:
Ah after more carefully reading the source code and document, I know the reason why retry_strategy.DEFAULT_BACKOFF_MAX = 10 is not working. When encounted an error,  a new Retry instance will be created by calling increment method of the current instance of Retry. So the change to the DEFAULT_BACKOFF_MAX will not survive after only one retry!
I will push a commit to fix it now.